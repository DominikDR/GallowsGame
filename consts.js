const GAME_STATE_FAILED = 'GAME_STATE_FAILED';
const GAME_STATE_WON = 'GAME_STATE_WON';
const MAX_ATTEMPTS = 6;
const GAME_STATE_KEYS = ['gameID', 'category', 'failsCounter', 'encodedPhrase', 'endState'];
const LETTER_STATUS_CORRECT = 'LETTER_STATUS_CORRECT';
const LETTER_STATUS_INCORRECT = 'LETTER_STATUS_INCORRECT';


module.exports = {
    GAME_STATE_FAILED,
    GAME_STATE_WON,
    MAX_ATTEMPTS,
    GAME_STATE_KEYS,
    LETTER_STATUS_CORRECT,
    LETTER_STATUS_INCORRECT,
};
