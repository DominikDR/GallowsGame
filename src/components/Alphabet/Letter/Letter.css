.letter {
    width: 80px;
    height: 80px;
    border: 1px solid #dedede;
    border-radius: 2px;
    color: #dedede;
    font-size: 30px;
    margin: 10px;
    cursor: pointer;
}