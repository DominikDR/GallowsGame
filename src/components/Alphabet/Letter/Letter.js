import React from 'react';
import styles from './Letter.css';
import classnames from 'classnames';

class Letter extends React.PureComponent {
    state = {
        letterStatus: '',
    };

    checkLetter = (letter, id) => {
        const url = `/phrases/check`;
        return fetch(url, {
            headers: {
                'Content-Type': 'application/json',
                'Accept': 'application/json',
            },
            method: 'post',
            body: JSON.stringify({
                id,
                letter,
            })
        })
        .then(response => response.json())
        .catch(error => {
            console.error(error);
        })
        .then(data => {
            this.setState({
                letterStatus: data.failsCounter === this.props.failsCounter ? 'LETTER_STATUS_CORRECT' : 'LETTER_STATUS_INCORRECT',
            })
            this.props.onLetterClick(data);
        });
    }
    
    render() {
        const { wasLetterUsed, letterStatus } = this.state;
        const { letter, gameID } = this.props;
        
        const letterStyle = classnames(styles.letter, {
            [styles.correctLetter]: letterStatus === 'LETTER_STATUS_CORRECT',
            [styles.wrongLetter]: letterStatus === 'LETTER_STATUS_INCORRECT',
        });
        return(
            <div className={letterStyle} onClick={this.checkLetter.bind(null, letter, gameID)}>
                {this.props.letter}
            </div>
        )
    }
}

export default Letter;