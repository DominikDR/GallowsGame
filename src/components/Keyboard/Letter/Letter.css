.letter {
    display: flex;
    align-items: center;
    justify-content: center;
    border: 1px solid #dedede;
    border-radius: 10px;
    color: #dedede;
    font-size: 22px;
    cursor: pointer; 
    outline: none;

    @media (min-width: 576px) {
        border-radius: 12px;
        font-size: 30px;
    }

    @media (min-width: 768px) {
        font-size: 35px;
    }
}

.letter:hover .letter:active {
    color: #191919;
    background-color: #dedede;
    border-color: #000;
}

.correctLetter {
    background-color: #157c00;
    pointer-events: none;
}

.wrongLetter {
    background-color: #bb0000;
    pointer-events: none;
}