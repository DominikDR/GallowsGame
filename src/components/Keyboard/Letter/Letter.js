import React from 'react';
import classnames from 'classnames';
import PropTypes from 'prop-types';
import styles from './Letter.css';
import { LETTER_STATUS_CORRECT, LETTER_STATUS_INCORRECT } from '../../../../consts';

class Letter extends React.PureComponent {
    checkLetter = async () => {
        const { letter, gameID, onLetterClick } = this.props;
        const url = '/phrases/check';
        try {
            const response = await fetch(url, {
                headers: {
                    'Content-Type': 'application/json',
                    Accept: 'application/json',
                },
                method: 'post',
                body: JSON.stringify({
                    gameID,
                    letter,
                }),
            });
            const data = await response.json();
            return onLetterClick(data, letter);
        } catch (error) {
            console.error(error);
            return error;
        }
    }

    render() {
        const { letter, letterStatus } = this.props;
        const letterStyle = classnames(styles.letter, {
            [styles.correctLetter]: letterStatus === LETTER_STATUS_CORRECT,
            [styles.wrongLetter]: letterStatus === LETTER_STATUS_INCORRECT,
        });
        return (
            <div className={letterStyle} onClick={this.checkLetter} tabIndex="0" role="button">
                {letter}
            </div>
        );
    }
}

Letter.propTypes = {
    letter: PropTypes.string.isRequired,
    gameID: PropTypes.number.isRequired,
    onLetterClick: PropTypes.func.isRequired,
    letterStatus: PropTypes.oneOfType([
        PropTypes.string,
    ]),
};

export default Letter;
