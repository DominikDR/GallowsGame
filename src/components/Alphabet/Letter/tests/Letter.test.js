import React from 'react';
import { shallow } from 'enzyme';
import Letter from '../Letter';
import { LETTER_STATUS_CORRECT, LETTER_STATUS_INCORRECT } from '../../../../../consts';

//1 method mount- mounts the whole tree of components
//2 shallow- do shallow render
// jest.fn() - creates anonymous function which always returns true

describe('<Letter />', () => {
    describe('Matches snapshot', () => {
        it('letter is neither correct nor incorrect', () => {
            const component = shallow(
                <Letter 
                    letter="D"
                    gameID={0}
                    onLetterClick={jest.fn()} 
                />
            );
            expect(component).toMatchSnapshot();
        });
        
        it('letter is correct', async () => {
            const component = shallow(
                <Letter 
                    letter="D"
                    gameID={0}
                    onLetterClick={jest.fn()} 
                />
            );
            await component.setState({ letterStatus: LETTER_STATUS_CORRECT });
            expect(component).toMatchSnapshot();
        });

        it('letter is incorrect', async () => {
            const component = shallow(
                <Letter 
                    letter="D"
                    gameID={0}
                    onLetterClick={jest.fn()} 
                />
            );
            await component.setState({ letterStatus: LETTER_STATUS_INCORRECT });
            expect(component).toMatchSnapshot();
        });

    })

})