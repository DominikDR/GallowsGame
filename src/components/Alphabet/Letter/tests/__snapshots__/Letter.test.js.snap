// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Letter /> Matches snapshot letter is correct 1`] = `
<div
  className="letter correctLetter"
  onClick={[Function]}
>
  D
</div>
`;

exports[`<Letter /> Matches snapshot letter is incorrect 1`] = `
<div
  className="letter wrongLetter"
  onClick={[Function]}
>
  D
</div>
`;

exports[`<Letter /> Matches snapshot letter is neither correct nor incorrect 1`] = `
<div
  className="letter"
  onClick={[Function]}
>
  D
</div>
`;
